### DISCUSSION TITLE
docs/g4noqasked/chap4_primgen/g4-noQasked-008-prigen/

### DISCUSSION THREAD

**Discussion by giscus** [evandde/blog_comment#62] [Announcements]
# docs/g4noqasked/chap4_primgen/g4-noQasked-008-prigen/

Geant4 무작정 따라하기 시리즈의 여덟번째. Event의 시작에 해당하는 선원항을 정의하는 방법에 대해 알아봅니다.

https://evanote.mew.kr/docs/g4noqasked/chap4_primgen/g4-noQasked-008-prigen/

**Comment by Fluence0115** [comment]
선생님 안녕하세요? 정성스럽게 작성하신 글 잘 보았습니다.

혹시 선원항이 heavy-ion(Li, Si, U 등)인 경우는 어떻게 적용해야 될까요?

아니면 참고할만한 example이 있을까요?

답변 주시면 감사하겠습니다.

**Reply by evandde** [reply]
1. Heavy Ion에 운동에너지를 주어 날리는 것이라면, 다음과 같은 형태로 정의할 수 있습니다.
``` C++
//fluorine
    G4int Z = 9, A = 18;
    G4double ionCharge   = 0.*eplus;
    G4double excitEnergy = 0.*keV;

    G4ParticleDefinition* ion
       = G4IonTable::GetIonTable()->GetIon(Z,A,excitEnergy);
    fParticleGun->SetParticleDefinition(ion);
    fParticleGun->SetParticleCharge(ionCharge);
```
Geant4가 동작하는 내부 구조 및 순서로 인해, geantino같은 다른 입자로 설정했다가 ion으로 바꿔줘야하는 번거로움이 있습니다만, 저 방법을 활용하시면 됩니다.
관련해서 Geant4 공식 예제인 `examples/basic/B3`의 PrimaryGeneratorAction 클래스를 참고하시기 바랍니다.

2-1. 이 강의에서 언급하지는 않았지만, macro 파일 (run.mac 같은 파일) 을 활용하여 G4ParticleGun의 선원 조건을 조정할 수도 있습니다.
```
/gun/particle ion
/gun/ion 92 238
/gun/energy 1 MeV
```
이런 식으로 말이죠. 

2-2. 마찬가지로 이 강의에서 언급하지는 않았지만, G4GeneralParticleSource라는 클래스를 활용하시면 좀 더 쉽게 작성할 수도 있습니다.

2-3. 다양한 선원항 정의방법에 대해서는 공식 예제인 examples/extended/eventgenerator를 참고해보시기 바랍니다.

3. 만약 방사성붕괴를 모사하셔야 한다면, 이 강의에서는 아예 언급하지 않는 "물리모델"이라는 부분을 추가하셔야합니다.
현재 제가 제공해드린 템플릿 코드는 QBBC 물리모델을 기반으로 하고 있습니다.
이는 방사성동위원소의 붕괴도식을 고려하지 않으므로, 예를 들어 238U를 놓는다해도 그 딸핵종으로 붕괴하는 일이 벌어지지 않습니다.
이를 고려하여 붕괴까지 시키려면 G4RadioactiveDecay라는 프로세스를 직접 추가하거나, 아니면 이를 포함하는 G4RadioactiveDecayPhysics와 같은 클래스가 추가되어야합니다.
관련해서는 공식 예제인 examples/extended/radioactivedecay를 참고해보시기 바랍니다.

**Reply by Fluence0115** [reply]
네 선생님 감사합니다. 적용해 보겠습니다~
